bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java

```diff
@@ -0,0 +1,210 @@
+    private @Nullable Customer lastData;
+    private int refresh;
+
+    public HydrawiseAccountHandler(final Bridge bridge, final HttpClient httpClient, final OAuthFactory oAuthFactory) {
+        super(bridge);
+        this.oAuthService = oAuthFactory.createOAuthClientService(getThing().toString(), AUTH_URL, AUTH_URL, CLIENT_ID,
+                CLIENT_SECRET, SCOPE, false);
+        oAuthService.addAccessTokenRefreshListener(this);
+        this.apiClient = new HydrawiseGraphQLClient(httpClient, oAuthService);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Handler initialized.");
+        scheduler.schedule(this::configure, 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Handler disposed.");
+        clearPolling();
+    }
+
+    @Override
+    public void onAccessTokenResponse(AccessTokenResponse tokenResponse) {
+        logger.debug("Auth Token Refreshed, expires in {}", tokenResponse.getExpiresIn());
+    }
+
+    public void addControllerListeners(HydrawiseControllerListener listener) {
+        this.controllerListeners.add(listener);
+        Customer data = lastData;
+        if (data != null) {
+            listener.onData(data.controllers);
+        }
+    }
+
+    public void removeControllerListeners(HydrawiseControllerListener listener) {
+        this.controllerListeners.remove(listener);
+    }
+
+    public @Nullable HydrawiseGraphQLClient graphQLClient() {
+        return apiClient;
+    }
+
+    public @Nullable Customer lastData() {
+        return lastData;
+    }
+
+    public void refreshData(int delaySeconds) {
+        initPolling(delaySeconds, this.refresh);
+    }
+
+    private void configure() {
+        HydrawiseAccountConfiguration config = getConfig().as(HydrawiseAccountConfiguration.class);
+        try {
+            if (!config.userName.isEmpty() && !config.password.isEmpty()) {
+                if (!config.savePassword) {
+                    Configuration editedConfig = editConfiguration();
+                    editedConfig.remove("password");
+                    updateConfiguration(editedConfig);
+                }
+                oAuthService.getAccessTokenByResourceOwnerPasswordCredentials(config.userName, config.password, SCOPE);
+            } else if (oAuthService.getAccessTokenResponse() == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Login credentials required.");
+                return;
+            }
+            this.refresh = Math.max(config.refreshInterval, MIN_REFRESH_SECONDS);
+            initPolling(0, refresh);
+        } catch (OAuthException | IOException e) {
+            logger.debug("Could not log in", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```

Comment:
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.